File: mpi-intro/minmeanmax-mpi.c
#include <stdio.h>
#include <mpi.h>
#include <stdlib.h>

int main(int argc, char **argv) {
    const int nx=1500;
    float *dat;
    int i;
    float datamin, datamax, datamean;
    float minmeanmax[3];
    float globminmeanmax[3];
    int ierr;
    int rank, size;
    int tag=1;
    int masterproc=0;
    MPI_Status status;


    ierr = MPI_Init(&argc, &argv);
    ierr = MPI_Comm_size(MPI_COMM_WORLD,&size);
    ierr = MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    /*
     * generate random data
     */

    dat = (float *)malloc(nx * sizeof(float));
    srand(0);
    for (i=0;i<nx;i++) {
        dat[i] = 2*((float)rand()/RAND_MAX)-1.;
    }

    /*
     * find min/mean/max
     */ 

    datamin = 1e+19;
    datamax =-1e+19;
    datamean = 0;
    
    for (i=0;i<nx;i++) {
        if (dat[i] < datamin) datamin=dat[i];
        if (dat[i] > datamax) datamax=dat[i];
        datamean += dat[i];
    }
    datamean /= nx;
    free(dat);

    minmeanmax[0] = datamin;
    minmeanmax[2] = datamax;
    minmeanmax[1] = datamean;

    if (rank != masterproc) {
       ierr = MPI_Ssend( /* ... ? ... */ );
    } else {
        globminmeanmax[0] = datamin;
        globminmeanmax[2] = datamax;
        globminmeanmax[1] = datamean;
        for (i=1;i<size;i++) {
            ierr = MPI_Recv( /* ... ? ... */);

            globminmeanmax[1] += minmeanmax[1];

            if (minmeanmax[0] < globminmeanmax[0])
                globminmeanmax[0] = minmeanmax[0];

            if (minmeanmax[2] > globminmeanmax[2])
                globminmeanmax[2] = minmeanmax[2];

        }
        globminmeanmax[1] /= size;
        printf("Min/mean/max = %f,%f,%f\n", globminmeanmax[0],
               globminmeanmax[1],globminmeanmax[2]);
    }
 
    ierr = MPI_Finalize();

    return 0;
}

File: mpi-intro/minmeanmax-mpi.f90
       program randomdata
       use mpi
       implicit none
       integer,parameter :: nx=1500
       real, allocatable :: dat(:)

       integer :: i,n
       real    :: datamin, datamax, datamean
       real    :: globmin, globmax, globmean
       integer :: ierr, rank, comsize
       integer :: ourtag=5
       real, dimension(3) :: sendbuffer, recvbuffer
       integer, dimension(MPI_STATUS_SIZE) :: status

       call MPI_INIT(ierr)
       call MPI_COMM_SIZE(MPI_COMM_WORLD, comsize, ierr)
       call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
!
! random data
!
       allocate(dat(nx))
       call random_seed(size=n)
       call random_seed(put=[(rank,i=1,n)])
       call random_number(dat)
       dat = 2*dat - 1.

!
! find min/mean/max
! 
       datamin = minval(dat)
       datamax = maxval(dat)
       datamean= (1.*sum(dat))/nx
       deallocate(dat)

       if (rank /= 0) then
            sendbuffer(1) = datamin
            sendbuffer(2) = datamean
            sendbuffer(3) = datamax
            call MPI_SSEND()  ! stuff goes here
       else
            globmin = datamin
            globmax = datamax
            globmean = datamean
            do i=2,comsize 
                call MPI_RECV() ! stuff goes here
                if (recvbuffer(1) < globmin) globmin=recvbuffer(1)
                if (recvbuffer(3) > globmax) globmax=recvbuffer(3)
                globmean = globmean + recvbuffer(2)
            enddo
            globmean = globmean / comsize
       endif

       print *,rank, ': min/mean/max = ', datamin, datamean, datamax

       if (rank==0) then
           print *, 'global min/mean/max = ', globmin, globmean, globmax
       endif 
  
       call MPI_FINALIZE(ierr)
       end
